flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/OverAggregateTestPrograms.java

```diff
@@ -0,0 +1,235 @@
+        "+I[Hello, 16, 1970-01-01T00:00:16, 16, 0, 3, 3]",
+        "+I[Hello, 20, 1970-01-01T00:00:20, 2, 0, 4, 5]",
+        "+I[Hello, 30, 1970-01-01T00:00:30, 3, 0, 5, 8]",
+        "+I[Hello, 40, 1970-01-01T00:00:40, 4, 0, 6, 12]",
+        "+I[Hello, 50, 1970-01-01T00:00:50, 5, 1, 7, 17]",
+        "+I[Hello, 60, 1970-01-01T00:01, 6, 2, 8, 23]",
+        "+I[Hello, 65, 1970-01-01T00:01:05, 65, 3, 9, 29]",
+        "+I[Hello, 90, 1970-01-01T00:01:30, 9, 4, 10, 35]",
+        "+I[Hello, 95, 1970-01-01T00:01:35, 18, 5, 11, 41]",
+        "+I[Hello World, 100, 1970-01-01T00:01:40, 7, 1, 1, 7]",
+        "+I[Hello World, 110, 1970-01-01T00:01:50, 17, 2, 2, 14]",
+        "+I[Hello World, 140, 1970-01-01T00:02:20, 18, 3, 3, 21]",
+        "+I[Hello World, 150, 1970-01-01T00:02:30, 8, 4, 4, 29]",
+        "+I[Hello World, 200, 1970-01-01T00:03:20, 20, 5, 5, 49]"
+    };
+
+    public static final String[] AFTER_RESTORE_DATA_UNBOUNDED = {
+        "+I[Hello World, 150, 1970-01-01T00:02:30, 8, 6, 6, 57]",
+        "+I[Hello World, 151, 1970-01-01T00:02:31, 8, 7, 7, 65]",
+        "+I[Hello World, 202, 1970-01-01T00:03:22, 20, 8, 8, 85]"
+    };
+
+    static final TableTestProgram OVER_AGGREGATE_UNBOUNDED_PARTITIONED_ROWS =
+            getTableTestProgram(
+                    "over-aggregate-unbounded-partitioned-rows",
+                    "validates over aggregate node",
+                    "PARTITION BY c ORDER BY rowtime RANGE UNBOUNDED PRECEDING",
+                    BEFORE_RESTORE_DATA_UNBOUNDED,
+                    AFTER_RESTORE_DATA_UNBOUNDED);
+
+    public static final String[] BEFORE_RESTORE_DATA_PRECEDING_ROWS = {
+        "+I[Hello, 10, 1970-01-01T00:00:10, 1, 0, 1, 1]",
+        "+I[Hello, 15, 1970-01-01T00:00:15, 15, 0, 2, 2]",
+        "+I[Hello, 16, 1970-01-01T00:00:16, 16, 0, 3, 3]",
+        "+I[Hello, 20, 1970-01-01T00:00:20, 2, 0, 4, 5]",
+        "+I[Hello, 30, 1970-01-01T00:00:30, 3, 0, 5, 8]",
+        "+I[Hello, 40, 1970-01-01T00:00:40, 4, 0, 6, 12]",
+        "+I[Hello, 50, 1970-01-01T00:00:50, 5, 1, 6, 16]",
+        "+I[Hello, 60, 1970-01-01T00:01, 6, 2, 6, 21]",
+        "+I[Hello, 65, 1970-01-01T00:01:05, 65, 3, 6, 26]",
+        "+I[Hello, 90, 1970-01-01T00:01:30, 9, 4, 6, 30]",
+        "+I[Hello, 95, 1970-01-01T00:01:35, 18, 5, 6, 33]",
+        "+I[Hello World, 100, 1970-01-01T00:01:40, 7, 1, 1, 7]",
+        "+I[Hello, 99, 1970-01-01T00:01:39, 15, 6, 6, 48]",
+        "+I[Hello World, 110, 1970-01-01T00:01:50, 17, 2, 2, 14]",
+        "+I[Hello World, 140, 1970-01-01T00:02:20, 18, 3, 3, 21]",
+        "+I[Hello World, 150, 1970-01-01T00:02:30, 8, 4, 4, 29]",
+        "+I[Hello World, 200, 1970-01-01T00:03:20, 20, 5, 5, 49]"
+    };
+
+    public static final String[] AFTER_RESTORE_DATA_PRECEDING_ROWS = {
+        "+I[Hello World, 202, 1970-01-01T00:03:22, 20, 6, 6, 69]"
+    };
+
+    static final TableTestProgram OVER_AGGREGATE_BOUNDED_PARTITIONED_PRECEDING_ROWS =
```

Comment:
Is `Preceding` differentiating enough? I believe the difference is `time-based` preceding vs `count-based` preceding.